### ISSUE TITLE
Layers with non-zero gradients are not updated with optimizer.apply_gradients()

### ISSUE THREAD

**Issue by QEDan** [tensorflow/tensorflow#45582]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Not tested
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):
Tested with both 2.2.0 and a recent 2.5.0 nightly build
v1.12.1-47149-g0939f0b7a8a 2.5.0-dev20201208
v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.7.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 10.1.243, CUDNN 7.6.5
- GPU model and memory: GTX 1070Ti

**Describe the current behavior**

In the following example, loading a second Keras model causes a situation where layers of the map_model are not updated with optimizer.apply_gradients() even though the gradients are non-zero. This is a silent failure, so I have added a callback called WeightChecker to demonstrate the issue. Otherwise, this code is meant to closely follow this (working) example: https://keras.io/examples/generative/dcgan_overriding_train_step/

```python
import numpy as np
import tensorflow as tf
from tensorflow import keras


class WeightChecker:
    """Automated health checks for training Keras models."""
    def __init__(self, model):
        self.initial_model = model
        self.var_names = [var.name for var in model.trainable_variables]
        self.prev_weights = model.get_weights()

    def check_epoch(self, model):
        """Checks to run at the end of an epoch"""
        self.check_untrained_params(model)

    def check_untrained_params(self, model):
        """Compare self.model.trainable_variables to self.prev_weights"""
        passed = True
        curr_weights = model.get_weights()
        for curr_var, prev_var, var_name in zip(curr_weights, self.prev_weights, self.var_names):
            eq = np.equal(curr_var, prev_var).all()
            if eq:
                passed = False
                print(f"\nWarning: Variable {var_name} was not updated with training. "
                      f"Confirm that this layer is correctly "
                      f"connected to the computation graph.")
        self.prev_weights = [w.copy() for w in curr_weights]
        return passed


class WeightCheckerCallback(keras.callbacks.Callback):
    """Check model initialization and run training checks.
    """
    def __init__(self):
        super().__init__()
        self.weight_check = None

    def setup_weight_checker(
            self,
            model: keras.Model = None):
        """Initialize the callback with an input_batch and targets."""
        self.weight_check = WeightChecker(model)

    def on_train_begin(self, logs=None):
        if self.weight_check is None:
            raise ValueError("setup_weight_checker() must be called to use WeightCheckerCallback.")

    def on_epoch_end(self, epoch, logs=None):
        self.weight_check.check_epoch(self.model)


def data_gen():
    """Generate random data for training."""
    data = (np.random.random((audio_len, 1)).astype(np.float32),
            np.random.random((audio_len, 1)).astype(np.float32))
    while True:
        yield data


batch_size = 64
audio_len = 16000
steps_per_epoch = 10
dataset = tf.data.Dataset.from_generator(data_gen,
                                         (tf.float32, tf.float32),
                                         (tf.TensorShape((audio_len, 1)), tf.TensorShape((audio_len, 1))))
dataset = dataset.batch(batch_size)

map_model = tf.keras.Sequential([tf.keras.layers.Conv1D(
        64, 3, padding='same'
    ),
    tf.keras.layers.Conv1D(
        1, 3, padding='same'
    )])
map_model(np.random.random((batch_size, audio_len, 1)))

aux_model = tf.keras.Sequential([tf.keras.layers.Dense(1)])
aux_model.trainable = False
aux_model(np.random.random((batch_size, 1)))


class MainModel(tf.keras.Model):
    """Main Model."""
    def __init__(self, map_model, aux_model):
        super().__init__()
        self.feature_dim = 128
        self.aux_model = aux_model
        self.map_model = map_model
        self.global_step = 0

    def call(self, inputs, training=True):
        output = self.map_model(inputs)
        return output

    def train_step(self, data):
        mixed_audio = data[0]
        clean_audio = data[1]

        with tf.GradientTape() as tape:
            decoded_audio = self.map_model(mixed_audio)
            total_loss = tf.reduce_mean(tf.abs(clean_audio - decoded_audio))

        grads = tape.gradient(total_loss, self.trainable_variables)
        [tf.print(f'Gradient std for {tv.name}: '
                  f'{np.std(g.numpy())}')
         for (g, tv) in zip(grads, self.trainable_variables)
         if 'conv1d/' in tv.name]
        self.optimizer.apply_gradients(zip(grads, self.trainable_variables))

        losses = {
            'loss': total_loss,
        }
        # tf.summary.scalar('loss', total_loss, step=self.global_step)
        self.global_step += 1
        return losses


epochs = 5

model = MainModel(map_model, aux_model)
model.compile(
    loss='mae',
    optimizer='adam',
    run_eagerly=True
)

weight_checker = WeightCheckerCallback()
weight_checker.setup_weight_checker(model)

model.fit(
    dataset,
    epochs=epochs,
    callbacks=[weight_checker],
    steps_per_epoch=steps_per_epoch
)
```

Running this code demonstrates that the gradients for conv1d are non-zero,
```
Gradient std for sequential/conv1d/kernel:0: 0.003920636139810085
Gradient std for sequential/conv1d/bias:0: 0.00040249430458061397
```
but the layer weights of conv1d are not updated
```
Warning: Variable sequential/conv1d/kernel:0 was not updated with training. Confirm that this layer is correctly connected to the computation graph.
Warning: Variable sequential/conv1d/bias:0 was not updated with training. Confirm that this layer is correctly connected to the computation graph.
```
In this case, the aux_model is not doing anything, it is only defined. However, the problem remains even if it is used in the computation.

**Describe the expected behavior**
The expected behaviour can be seen by setting `aux_model = None` in the above code. When this is done, the conv1d layer updates as expected.

**Standalone code to reproduce the issue**
See problem description

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2, TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/253fc483f128e86c9b6dc3f3b8f886c8/45582-tf-nightly.ipynb). Thanks!

**Comment by WindQAQ** [comment]
Hi @QEDan, I think it's because `get_weights` returns all weights including trainable and non-trainable. You can validate this by printing the length of `trainable_variables` and `get_weights`. Therefore, your weight checker cannot pass because it examines the non-trainable variables. To fix it, just extract trainable variables only in the weight checker.

https://colab.research.google.com/drive/1ytXiosdXa-4SMCRETVcstq4ubS0GRU84?usp=sharing

**Comment by QEDan** [comment]
Ah, you are right. Thanks for looking into this. I'm sorry for the false alarm.
